### ISSUE TITLE
bug: false positive no-derived-state using resiseObserver

### ISSUE THREAD

**Issue by nirtamir2** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#24]
### Did you check docs and existing issues?

- [x] I have read all the plugin docs
- [x] I have searched the existing issues

### Plugin version

0.4.1

### ESLint version

9.0.0

### Describe the bug

```ts
function useHasOverflow<T extends HTMLElement>({
  contentRef,
  maxHeight,
}: {
  contentRef: RefObject<T | null>;
  maxHeight: number;
}) {
  const [hasOverflow, setHasOverflow] = useState(false);
  useEffect(() => {
    function checkOverflow() {
      if (contentRef.current != null) {
        const hasContentOverflow = contentRef.current.scrollHeight > maxHeight;
        // eslint-disable-next-line react-you-might-not-need-an-effect/no-derived-state
        setHasOverflow(hasContentOverflow);
      }
    }

    checkOverflow();

    const resizeObserver = new ResizeObserver(() => {
      checkOverflow();
    });

    if (contentRef.current != null) {
      resizeObserver.observe(contentRef.current);
    }

    return () => {
      resizeObserver.disconnect();
    };
  }, [contentRef, maxHeight]);

  return hasOverflow;
}
```

I get an error on         setHasOverflow(hasContentOverflow) but its should be fine

### Steps To Reproduce

Write this hook
See no error

### Expected Behavior

No error

**Comment by NickvanDyke** [comment]
I think you're correct, the plugin should not consider refs to be internal state. I'll push a fix when I get the time!

**Comment by NickvanDyke** [comment]
I think you're correct - the plugin should not consider refs to be internal state. However that may be difficult to reliably detect here because the ref comes from a prop, not a local `useRef`. At best I might be able to check the variable name for `ref`.

**Comment by NickvanDyke** [comment]
Actually looking at the [ResizeObserver docs](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver), I think a more correct (and readable) use of its API would fix this too. You should use the arg passed to `ResizeObserver`'s function instead of referencing closures (`contentRef`). The plugin knows the function is probably a callback and thus won't analyze setters inside it.

Here's a tweaked version that isn't flagged in tests:

```js
          function useHasOverflow({ contentRef, maxHeight }) {
            const [hasOverflow, setHasOverflow] = useState(false);

            useEffect(() => {
              const resizeObserver = new ResizeObserver((element) => {
                const hasContentOverflow = element.scrollHeight > maxHeight;
                setHasOverflow(hasContentOverflow);
              })

              if (contentRef.current != null) {
                resizeObserver.observe(contentRef.current);
              }

              return () => {
                resizeObserver.disconnect();
              };
            }, [contentRef, maxHeight]);

            return hasOverflow;
          }

```

Let me know if that functions the same as your example 🙂
